### PR TITLE
#695 user control over the mapping process

### PR DESCRIPTION
**Pull request by sjaakd**
Status of PR: Ready for review.

- [x] API
- [x] Basic Implementation
- [x] Documentation
- [x] Unit Test

Cleanup in setting of `xyzOption` to `xyzMapping` needs to enhanced. Initially I thought of adding this to this PR.  However, I'll handle it separately, this makes it more transparent.   So, no cleanup here.

- [ ] Cleanup

### CODE REVIEW

**Comment by sjaakd** [comment]
> * Are we leaking to much of the internal implementation into the API and the mapping control (perhaps it is OK since it is explained in the documentation).

1. The funny thing is, it was already there in the documentation. I just numbered the lines and aligned them with the implementation.

> * Maybe we need to add code examples in the different mapping control in the example (to show what it means)

2. Good point... I'll add some..

> * Do we really need annotations for this? I've been playing with some annotations recently and perhaps we can make it with an enum.

3. I think we can. But would it make the interface more extendable / usable? It is more verbose. I like the idea of meta-annotations(we introduced them for `@Mapping` recently  and thought we were going in that direction. BTW: thinking out loud, that might be the answer on the point I'm trying to make below at point 4..

> I am also not sure whether using `Class<? extends Annotation> mappingControl() default DefaultControl.class;` would make it easy for the users to detect this. If we allow meta annotations, they should just be able to do `@DeepClone` on their mapper (or method).

4. I don't like it either.  Somewhere early December (when we discussed this first time around on gitter) I had exactly this an an initial implementation. . But, you can't control an individual mapping anymore, something that you definitely would want as a user. Mixing them (so have them as meta annotation on places were we would also have a `@Mapper`, `@MapperConfig`, `@BeanMapping` and `@IterableMapping` / `@MapMapping` but deviating for `@Mapping` would not be consistent / confusing either in the inheritance we defined. Also... qualifiers work in the same way, and they are used quite extensively. 

But.. while answering that.. Since we have  a meta annotation on `@Mapping` we might use that to carry a meta-annotation on control as well. Although.. it feels pretty bulky:

So something like this..

instead of (or perhaps both):

```java

public interface MyMapper{

    @Mapping( target = "person", mappingControl = DeepClone.class )
     Entity map(Dto in);

}

```

We would have:

```java

public interface MyMapper{

    @DeepClonePerson
     Entity map(Dto in);

}

@Retention(RetentionPolicy.CLASS)
@DeepClone
@Mapping(target = "person")
 public @interface DeepClonePerson { }
```

The nice thing is.. we can bolt this on in a follow on PR allowing both options, not only for mappingControl, but also for other fields of `@Mapping`.. But we need to consider how obvious inheriting stuff from other levels (`@Mapper`, `@MappingConfig`) is


---
Summarising.. 
1. I'm still in favour of individual annotations instead of an enum, allowing a way to the future.
2. Also leave the `Class<? extends Annotation> mappingControl() default DefaultControl.class;`
3. But investigate how to extend on the idea of meta-annotations and especially, how it fits in to the newly introduced chain-of-responsibility / inheritance-hierarchy pattern.

File: core/src/main/java/org/mapstruct/MapMapping.java [hunk 1]
```diff
@@ -167,4 +169,22 @@
...
+     * Allows detailed control over the mapping process.
+     *
+     * @return the mapping control
+     *
+     * @since 1.4
+     */
+    Class<? extends Annotation> keyMappingControl() default DefaultControl.class;
+
+    /**
+     * Allows detailed control over the mapping process.
```
[the remark this answers is not in the document]
Comment (sjaakd) [reply core/src/main/java/org/mapstruct/MapMapping.java]: done

File: core/src/main/java/org/mapstruct/control/AllowBuiltInConversion.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import org.mapstruct.util.Experimental;
+
+/**
+ * Controls the mapping, allows for type conversion from source type to target type
+ *
+ * Type conversions are typically suppored directly in Java. The "toString()" is such an example,
+ * which allows for mapping for instance a  {@link java.lang.Number}type to a {@link java.lang.String}.
```
[the remark this answers is not in the document]
Comment (sjaakd) [reply core/src/main/java/org/mapstruct/control/AllowBuiltInConversion.java]: done

File: core/src/main/java/org/mapstruct/BeanMapping.java [hunk 3]
```diff
@@ -156,4 +158,14 @@ NullValuePropertyMappingStrategy nullValuePropertyMappingStrategy()
...
     Builder builder() default @Builder;
+
+    /**
+     * Allows detailed control over the mapping process.
+     *
+     * @return the mapping control
+     *
+     * @since 1.4
+     */
+    Class<? extends Annotation> mappingControl() default DefaultControl.class;
```
[the remark this answers is not in the document]
Comment (sjaakd) [reply core/src/main/java/org/mapstruct/BeanMapping.java]: done

File: processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java [hunk 4]
```diff
@@ -206,47 +210,52 @@ private Assignment getTargetAssignment(Type sourceType, Type targetType) {
...
-            }
 
-            // check for a built-in method
-            if (!hasQualfiers() ) {
-                Assignment builtInMethod = resolveViaBuiltInMethod( sourceType, targetType );
-                if ( builtInMethod != null ) {
-                    builtInMethod.setAssignment( sourceRHS );
-                    usedSupportedMappings.addAll( supportingMethodCandidates );
-                    return builtInMethod;
+                // TODO: should we distinguish builtIn from typeConversion, what does the documentation say?
```
[the remark this answers is not in the document]
Comment (sjaakd) [reply processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java]: ah.. but we never distinguish between the mechanism of typeconversion and built-in conversion in the documentation. And I think that's good. It is not of the users concern whether we do an in-lining of generate a method..

File: processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java [hunk 5]
```diff
@@ -270,6 +280,26 @@ private boolean hasQualfiers() {
             return selectionCriteria != null && selectionCriteria.hasQualfiers();
         }
 
+        private boolean allowDirect( Type sourceType, Type targetType ) {
```
[the remark this answers is not in the document]
Comment (sjaakd) [reply processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java]: good point. I'll change that. So all that starts with `java`?

File: processor/src/test/java/org/mapstruct/ap/test/mappingcontrol/MappingControlTest.java [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    @Test
+    @WithClasses(DirectMapper.class)
+    public void directSelectionAllowed() {
+
+        FridgeDTO in = createFridgeDTO();
+        FridgeDTO out = DirectMapper.INSTANCE.map( in );
+
+        assertThat( out ).isNotNull();
+        assertThat( out.getShelve() ).isNotNull();
+        assertThat( out.getShelve() ).isEqualTo( in.getShelve() );
```
[the remark this answers is not in the document]
Comment (sjaakd) [reply processor/src/test/java/org/mapstruct/ap/test/mappingcontrol/MappingControlTest.java]: done

File: processor/src/test/java/org/mapstruct/ap/test/mappingcontrol/MappingControlTest.java [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    @WithClasses(CloningMapper.class)
+    public void testDeepCloning() {
+
+        FridgeDTO in = createFridgeDTO();
+        FridgeDTO out = CloningMapper.INSTANCE.clone( in );
+
+        assertThat( out ).isNotNull();
+        assertThat( out.getShelve() ).isNotNull();
+        assertThat( out.getShelve() ).isNotEqualTo( in.getShelve() );
+        assertThat( out.getShelve().getCoolBeer() ).isNotEqualTo( in.getShelve().getCoolBeer() );
```
[the remark this answers is not in the document]
Comment (sjaakd) [reply processor/src/test/java/org/mapstruct/ap/test/mappingcontrol/MappingControlTest.java]: done

File: processor/src/test/java/org/mapstruct/ap/test/mappingcontrol/MappingControlTest.java [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+    @Test
+    @WithClasses(CloningMapper.class)
+    public void testDeepCloning() {
+
+        FridgeDTO in = createFridgeDTO();
+        FridgeDTO out = CloningMapper.INSTANCE.clone( in );
+
+        assertThat( out ).isNotNull();
+        assertThat( out.getShelve() ).isNotNull();
+        assertThat( out.getShelve() ).isNotEqualTo( in.getShelve() );
```
[the remark this answers is not in the document]
Comment (sjaakd) [reply processor/src/test/java/org/mapstruct/ap/test/mappingcontrol/MappingControlTest.java]: done

File: processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (filiphr) [reply processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java]: What I meant is that we should not ignore the built in methods when mapping methods are ignored. You are right the way we do the conversion is not the users concern

File: processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (filiphr) [reply processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java]: I'd say yes

**Comment by filiphr** [comment]
Thanks for the feedback and changes.

>1. The funny thing is, it was already there in the documentation. I just numbered the lines and aligned them with the implementation.

Yes you are right. I had the same feeling when I was reading your documentation changes. 

Regarding 3 and 4. My idea is still to allow meta annotations. However, instead of having 3 / 4 annotations that are like enums we can have 1 annotation for mapping control with array of enums. It is still future proof and people can still compose their own annotations.

As for supporting mapping control on a mapping bases then indeed we would have to do it in the `@Mapping`. The proposal with meta annotating your own mapping will not work. Since that meta mapping is applied on the method level not the property level. At least that's how I would expect meta annotating to work. You are applying your custom annotation to the element you are annotating.

File: processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (sjaakd) [reply processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java]: ok.. I removed the TODO.. type conversion == built-in conversion == handled by same control `AllowBuiltInConversion`.

File: documentation/src/main/asciidoc/chapter-5-data-type-conversions.asciidoc [hunk 9]
```diff
@@ -145,15 +145,31 @@ That way it is possible to map arbitrary deep object graphs. When mapping from e
...
 
-* If source and target attribute have the same type, the value will be simply copied from source to target. If the attribute is a collection (e.g. a `List`) a copy of the collection will be set into the target attribute.
-* If source and target attribute type differ, check whether there is another mapping method which has the type of the source attribute as parameter type and the type of the target attribute as return type. If such a method exists it will be invoked in the generated mapping implementation.
-* If no such method exists MapStruct will look whether a built-in conversion for the source and target type of the attribute exists. If this is the case, the generated mapping code will apply this conversion.
-* If no such method was found MapStruct will try to generate an automatic sub-mapping method that will do the mapping between the source and target attributes.
-* If MapStruct could not create a name based mapping method an error will be raised at build time, indicating the non-mappable attribute and its path.
+["arabic", start=1]
```
[the remark this answers is not in the document]
Comment (sjaakd) [reply documentation/src/main/asciidoc/chapter-5-data-type-conversions.asciidoc]: fixed

File: processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (sjaakd) [reply processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java]: done..

**Comment by sjaakd** [comment]
> The proposal with meta annotating your own mapping will not work. Since that meta mapping is applied on the method level not the property level. At least that's how I would expect meta annotating to work. You are applying your custom annotation to the element you are annotating.

I think it will. Have a look at this example:

```java
public interface MyMapper{

    @Mapping( target = "person", mappingControl = DeepClone.class )
    @Mapping( target = "pet", mappingControl = NoComplexMapping.class )
     Entity map(Dto in);

}
```


```java
public interface MyMapper{

public interface MyMapper{

    @DeepClonePerson
    @NoComplexPet
     Entity map(Dto in);

}

@Retention(RetentionPolicy.CLASS)
@DeepClone
@Mapping(target = "person")
 public @interface DeepClonePerson { }

@Retention(RetentionPolicy.CLASS)
@ NoComplexMapping.class
@Mapping(target = "pet")
 public @interface NoComplexPet { }


}
```
Its only a bit more cumbersome.. But remember: you can already compose mappings in the current code base.

**Comment by sjaakd** [comment]
> 1 annotation for mapping control with array of enums

- What about duplicates? We need to built in extra warnings? With annotations uniqueness comes for free
- What about inheritance trough the `@MapperConfig`, `@Mapper`.. etc? You will obfuscate this because you can define your meta's on another place

I would like to go for this approach and bolt-on the meta's later (including all other properties that we have now).. We could even (quite easily) scan an annotation for meta's..

**Comment by filiphr** [comment]
The first example is correct, the second one the `@NoComplexMapping` and `@DeepClone` are applied to the method. At least that's how a user would think. To me that is equivalent to writing:

```
public interface MyMapper{

    @NoComplexMapping
    @DeepClone
    @Mapping(target = "pet")
    @Mapping(target = "person")
    Entity map(Dto in);

}
```

If we start to handle this differently, users would be surprised and it is not something that is expected from meta annotations to work.

> * What about duplicates? We need to built in extra warnings? With annotations uniqueness comes for free

Well if we put them in a set in the processing it is OK. We don't care about duplicates. It doesn't hurt to write them. We would use an `EnumSet`.

> * What about inheritance trough the @MapperConfig, @Mapper.. etc? You will obfuscate this because you can define your meta's on another place

The proposal is not to have inheritence via `@MapperConfig` and `@Mapper`. Otherwise it is conflicting. Anyways we can also have `MappingControl.Step` in the `@MapperConfig` that way and people will have auto complete support.

Inheritance would be the same. If you find some `@MappingControl` on the method that has precedence, then `@BeanMapping`, then annotation on the interface, then `@Mapper`, then `@MapperConfig`.

**Comment by sjaakd** [comment]
Ok. lets park the discussion on the enums.. I can live with that (although its some work to change the current solution). 

And.. I see your point on the `@Mapping` meta-annotation.. That will become troublesome..

But.. I would like to tackle the meta-annotations in a a more general / complete way.. Not part of this issue.. We could even drop the current wow in 2.0 in favour for meta-annotations , if we introduce them in time.. (e.g. 1.4)

**Comment by sjaakd** [comment]
is superseded by #2015

**Pull request closed without merging**